docs/development/best-practices.md

```diff
@@ -0,0 +1,139 @@
+# Best practices
+
+This document explains our best practices.
+Follow these best practices when you're working on our code.
+
+## General
+
+- Prefer full function declarations for readability and better stack traces, so avoid `const func = ():void => {}`
+- Avoid [Enums](https://github.com/renovatebot/renovate/issues/13743), use union or [immutable objects](https://github.com/renovatebot/renovate/blob/5043379847818ac1fa71ff69c098451975e95710/lib/modules/versioning/pep440/range.ts#L8-L20) instead
+- Always add unit tests for full code coverage
+  - Only use `istanbul` comments for unreachable code coverage that is needed for `codecov` completion
+  - Use descriptive `istanbul` comments
+
+```js
+// istanbul ignore next: can never happen
+```
+
+### Logging
+
+Use logger metadata if logging for `WARN`, `ERROR`, `FATAL`, or if the result is a complex metadata object needing a multiple-line pretty stringification.
+Otherwise, inline metadata into the log message if logging at `INFO` or below, or if the metadata object is complex.
+
+`WARN` and above messages are often used in metrics or error catching services, and should have a consistent `msg` component so that they will be automatically grouped/associated together.
+Metadata which is separate from its message is harder for human readability, so try to combine it in the message unless it's too complex to do so.
+
+Good:
+
+```js
+logger.debug({ config }, 'Full config');
+logger.warn({ presetName }, 'Failed to look up preset');
+```
+
+Avoid:
+
+```js
+logger.debug({ branchName }, 'Generated branchName');
```

Comment:
🤔 It's not clear to me why this is bad and the `debug` above is good?

Reply:
Because it's just a string, which can be easily inlined. Otherwise it will take up 4 lines of pretty output for no reason. Example:

![image](https://user-images.githubusercontent.com/6311784/169286176-70420a05-6475-47dd-acc2-24b63eb7b4f2.png)

Apart from the fact that the log message is factually incorrect (it actually is the default endpoint..) that log would be better as one line and not 4.